# Datetime in Pandas

In [1]:
import pandas as pd
import datetime as dt

from pandas_datareader import data

Let's see how a date is described in Python

In [2]:
date = dt.date(1986, 12, 27)       # year, month, date

print(date)
print(type(date))

print('-'*20)

print(date.year)
print(date.month)
print(date.day)

1986-12-27
<class 'datetime.date'>
--------------------
1986
12
27


In [3]:
date = dt.date(1986, 12, 27)
date                # just to check the difference between direct printing and using print

datetime.date(1986, 12, 27)

Now, let's see how a date and time is expressed in Python

In [5]:
date_time = dt.datetime(1986, 12, 27, 17, 30, 20)     # year, month, date, hour, minutes, second

print(date_time)
print(type(date_time))

print('-'*20)

print(date_time.year)
print(date_time.month)
print(date_time.day)

print('-'*20)

print(date_time.hour)
print(date_time.minute)
print(date_time.second)

1986-12-27 17:30:20
<class 'datetime.datetime'>
--------------------
1986
12
27
--------------------
17
30
20


**Timestamp**

In [6]:
date = pd.Timestamp("1986 12 27")
date

Timestamp('1986-12-27 00:00:00')

In [7]:
date = pd.Timestamp("1986/12/27")      # expressed exactly as the above representation
date

Timestamp('1986-12-27 00:00:00')

In [8]:
date = pd.Timestamp("1986, 12, 27")    # works the same
date

Timestamp('1986-12-27 00:00:00')

We see that Timestamp is quite useful, parsing dates written in various formats into a standard format. Now, let's see how to include time in the Timestamp representation

In [9]:
datetime = pd.Timestamp("1986-12-27 17:30:20")
datetime

Timestamp('1986-12-27 17:30:20')

In [10]:
datetime = pd.Timestamp("1986-12-27 5:30:20 PM")    # same as above
datetime

Timestamp('1986-12-27 17:30:20')

In [11]:
datetime = pd.Timestamp("1986-12-27 5:30:20 AM")
datetime

Timestamp('1986-12-27 05:30:20')

In [12]:
datetime = pd.Timestamp("1986-12-27 5:30:20")
datetime

Timestamp('1986-12-27 05:30:20')

**DateTimeIndex**

We create a list of strings and then convert to DatetimeIndex

In [13]:
dates = ["1986/12/27", "1986/12/28", "1986/12/29"]  # create a list of strings

pd.DatetimeIndex(dates)    # converting to a real time datetime index

DatetimeIndex(['1986-12-27', '1986-12-28', '1986-12-29'], dtype='datetime64[ns]', freq=None)

We can also create the the list of dates using datetime and then convert the list to the DatetimeIndex

In [14]:
dates = [dt.date(1986, 12, 27), dt.date(1986, 12, 28), dt.date(1986, 12, 29)]
pd.DatetimeIndex(dates)

DatetimeIndex(['1986-12-27', '1986-12-28', '1986-12-29'], dtype='datetime64[ns]', freq=None)

Now, lets assign some values to these dates

In [15]:
dates = ["1986/12/27", "1986/12/28", "1986/12/29"]
date_index = pd.DatetimeIndex(dates)

values = ('Sudhir', 'Narayan', 'Pathak')
pd.Series(data = values, index = date_index)

1986-12-27     Sudhir
1986-12-28    Narayan
1986-12-29     Pathak
dtype: object

**pd.to_datetime() method**

In [16]:
date = pd.to_datetime('1986, 12, 27')  # string input
date 

Timestamp('1986-12-27 00:00:00')

In [17]:
date = pd.to_datetime('1986/12/27')     # string input
date

Timestamp('1986-12-27 00:00:00')

In [18]:
date = pd.to_datetime(dt.date(1986, 12, 27))   # datetime input, see below example
date

Timestamp('1986-12-27 00:00:00')

In [19]:
date = dt.date(1986, 12, 27)     # this is of datetime type
date                             # calling to_datetime on this converts it to Timestamp, as above 

datetime.date(1986, 12, 27)

In [20]:
datetime = pd.to_datetime(dt.datetime(1986, 12, 27, 17, 30, 20))
datetime

Timestamp('1986-12-27 17:30:20')

In [21]:
dates = ["1986/12/27", "1986/12/28", "1986/12/29"]   # list of strings
pd.to_datetime(dates)     # to_datetime generates a DateTimeIndex, instead of a Timestamp

DatetimeIndex(['1986-12-27', '1986-12-28', '1986-12-29'], dtype='datetime64[ns]', freq=None)

pd.to_datetime() can be used to convert the columns of series or dataframe to a standard datetime format. We can see how different formats have been converted to the standard datetime format.

In [22]:
dates = ['December 22nd, 2019', '2019, 12, 22', 'This is Date', 'Dec 23rd, 2019', '24th Dec 2019']
pd.to_datetime(dates, errors = 'coerce')

DatetimeIndex(['2019-12-22', '2019-12-22', 'NaT', '2019-12-23', '2019-12-24'], dtype='datetime64[ns]', freq=None)

**pd.date_range**

In [23]:
dates = pd.date_range(start = '2019-12-22', end = '2020-12-22', freq = 'D')
dates

DatetimeIndex(['2019-12-22', '2019-12-23', '2019-12-24', '2019-12-25',
               '2019-12-26', '2019-12-27', '2019-12-28', '2019-12-29',
               '2019-12-30', '2019-12-31',
               ...
               '2020-12-13', '2020-12-14', '2020-12-15', '2020-12-16',
               '2020-12-17', '2020-12-18', '2020-12-19', '2020-12-20',
               '2020-12-21', '2020-12-22'],
              dtype='datetime64[ns]', length=367, freq='D')

It has created a total of 367 values, inclusive of both end points. Now, let's change the frequency to Annualy

In [24]:
dates = pd.date_range(start = '2019-12-22', end = '2020-12-22', freq = 'A')
dates

DatetimeIndex(['2019-12-31'], dtype='datetime64[ns]', freq='A-DEC')

In [25]:
dates = pd.date_range(start = '2019-12-22', end = '2022-12-22', freq = 'A')
dates

DatetimeIndex(['2019-12-31', '2020-12-31', '2021-12-31'], dtype='datetime64[ns]', freq='A-DEC')

Now, lets generate it weekely

In [26]:
dates = pd.date_range(start = '2019-12-22', end = '2022-12-22', freq = 'W')
dates

DatetimeIndex(['2019-12-22', '2019-12-29', '2020-01-05', '2020-01-12',
               '2020-01-19', '2020-01-26', '2020-02-02', '2020-02-09',
               '2020-02-16', '2020-02-23',
               ...
               '2022-10-16', '2022-10-23', '2022-10-30', '2022-11-06',
               '2022-11-13', '2022-11-20', '2022-11-27', '2022-12-04',
               '2022-12-11', '2022-12-18'],
              dtype='datetime64[ns]', length=157, freq='W-SUN')

In [27]:
dates = pd.date_range(start = '2019-12-22', end = '2022-12-22', freq = '4W')
dates

DatetimeIndex(['2019-12-22', '2020-01-19', '2020-02-16', '2020-03-15',
               '2020-04-12', '2020-05-10', '2020-06-07', '2020-07-05',
               '2020-08-02', '2020-08-30', '2020-09-27', '2020-10-25',
               '2020-11-22', '2020-12-20', '2021-01-17', '2021-02-14',
               '2021-03-14', '2021-04-11', '2021-05-09', '2021-06-06',
               '2021-07-04', '2021-08-01', '2021-08-29', '2021-09-26',
               '2021-10-24', '2021-11-21', '2021-12-19', '2022-01-16',
               '2022-02-13', '2022-03-13', '2022-04-10', '2022-05-08',
               '2022-06-05', '2022-07-03', '2022-07-31', '2022-08-28',
               '2022-09-25', '2022-10-23', '2022-11-20', '2022-12-18'],
              dtype='datetime64[ns]', freq='4W-SUN')

Similarly to 4W gap, we can try 2D, 3D gaps in above examples. We can also generate it hourly

In [29]:
times = pd.date_range(start = '2019-12-22', end = '2019-12-30', freq = '6H')
times

DatetimeIndex(['2019-12-22 00:00:00', '2019-12-22 06:00:00',
               '2019-12-22 12:00:00', '2019-12-22 18:00:00',
               '2019-12-23 00:00:00', '2019-12-23 06:00:00',
               '2019-12-23 12:00:00', '2019-12-23 18:00:00',
               '2019-12-24 00:00:00', '2019-12-24 06:00:00',
               '2019-12-24 12:00:00', '2019-12-24 18:00:00',
               '2019-12-25 00:00:00', '2019-12-25 06:00:00',
               '2019-12-25 12:00:00', '2019-12-25 18:00:00',
               '2019-12-26 00:00:00', '2019-12-26 06:00:00',
               '2019-12-26 12:00:00', '2019-12-26 18:00:00',
               '2019-12-27 00:00:00', '2019-12-27 06:00:00',
               '2019-12-27 12:00:00', '2019-12-27 18:00:00',
               '2019-12-28 00:00:00', '2019-12-28 06:00:00',
               '2019-12-28 12:00:00', '2019-12-28 18:00:00',
               '2019-12-29 00:00:00', '2019-12-29 06:00:00',
               '2019-12-29 12:00:00', '2019-12-29 18:00:00',
               '2019-12-

**datareader**

In [30]:
company = 'MSFT'             # Microsoft
start_date = '2019-01-01'
end_date = '2019-12-22'

stocks = data.DataReader(name = company,
                         data_source = 'yahoo',   # google finance provides the data
                         start = start_date,
                         end = end_date)

stocks.shape

(247, 6)

In [31]:
stocks             # data is only for working days

High         Low        Open       Close      Volume  \
Date                                                                     
2018-12-31  102.400002  100.440002  101.290001  101.570000  33173800.0   
2019-01-02  101.750000   98.940002   99.550003  101.120003  35329300.0   
2019-01-03  100.190002   97.199997  100.099998   97.400002  42579100.0   
2019-01-04  102.510002   98.930000   99.720001  101.930000  44060600.0   
2019-01-07  103.269997  100.980003  101.639999  102.059998  35656100.0   
...                ...         ...         ...         ...         ...   
2019-12-16  155.899994  154.820007  155.110001  155.529999  24144200.0   
2019-12-17  155.710007  154.449997  155.449997  154.690002  25425600.0   
2019-12-18  155.479996  154.179993  154.300003  154.369995  24129200.0   
2019-12-19  155.770004  153.750000  154.000000  155.710007  24958900.0   
2019-12-20  158.490005  156.289993  157.350006  157.410004  53477500.0   

             Adj Close  
Date                    
2018-12-31   99.817421  
2019-01-02   99.375191  
2019-01-03   95.719376  
2019-01-04  100.171211  
2019-01-07  100.298965  
...                ...  
2019-12-16  155.106339  
2019-12-17  154.268631  
2019-12-18  153.949509  
2019-12-19  155.285858  
2019-12-20  156.981232  

[247 rows x 6 columns]

In [33]:
stocks.loc['2019-01-02']    # extracting a particular row

High         1.017500e+02
Low          9.894000e+01
Open         9.955000e+01
Close        1.011200e+02
Volume       3.532930e+07
Adj Close    9.937519e+01
Name: 2019-01-02 00:00:00, dtype: float64

In [34]:
stocks.iloc[1]              # extracting using the row index

High         1.017500e+02
Low          9.894000e+01
Open         9.955000e+01
Close        1.011200e+02
Volume       3.532930e+07
Adj Close    9.937519e+01
Name: 2019-01-02 00:00:00, dtype: float64

In [37]:
stocks.loc['2019-12-2' : '2019-12-16']    # extracting a range of dates, both ends inclusive

High         Low        Open       Close      Volume  \
Date                                                                     
2019-12-02  151.830002  148.320007  151.809998  149.550003  27418400.0   
2019-12-03  149.429993  146.649994  147.490005  149.309998  24066000.0   
2019-12-04  150.179993  149.199997  150.139999  149.850006  17574700.0   
2019-12-05  150.320007  149.479996  150.050003  149.929993  17869100.0   
2019-12-06  151.869995  150.270004  150.990005  151.750000  16403500.0   
2019-12-09  152.210007  150.910004  151.070007  151.360001  16687400.0   
2019-12-10  151.889999  150.759995  151.289993  151.130005  16476100.0   
2019-12-11  151.869995  150.330002  151.539993  151.699997  18856600.0   
2019-12-12  153.440002  151.020004  151.649994  153.240005  24612100.0   
2019-12-13  154.889999  152.830002  153.000000  154.529999  23845400.0   
2019-12-16  155.899994  154.820007  155.110001  155.529999  24144200.0   

             Adj Close  
Date                    
2019-12-02  149.142639  
2019-12-03  148.903290  
2019-12-04  149.441818  
2019-12-05  149.521591  
2019-12-06  151.336639  
2019-12-09  150.947708  
2019-12-10  150.718338  
2019-12-11  151.286774  
2019-12-12  152.822586  
2019-12-13  154.109070  
2019-12-16  155.106339

We can truncate rows before and after a particular date

In [ ]:
stocks.truncate(before = '2019-12-2', after = '2019-12-16')

**Timedelta**

In [ ]:
time_a = pd.Timestamp('2019-12-22 17:23:34')
time_b = pd.Timestamp('2019-12-23 19:03:14')

time_b - time_a

In [ ]:
pd.Timedelta(weeks = 10, days = 2, hours = 3, minutes = 23)

In [ ]:
df = pd.read_csv('ecommerce.csv', index_col = 'ID')
df.shape

In [ ]:
df

In [ ]:
df.dtypes

As we can see, Python automaticaly does not read the dates in datetime format. We need to parse it as follows:

In [ ]:
df = pd.read_csv('ecommerce.csv', 
                 index_col = 'ID', 
                 parse_dates = ['order_date', 'delivery_date'])   # passing the columns to parse
df.head()

In [ ]:
df.dtypes

In [ ]:
df['Delivery Time'] = df['delivery_date'] - df['order_date']
df.head()